java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -0,0 +1,108 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class SymbolicValueRelation {
+
+  @Nullable
+  public static Boolean checkValidity(SymbolicValueRelation relation, Collection<SymbolicValueRelation> knownRelations) {
+    Boolean result = checkValidityNoInverse(relation, knownRelations);
+    if (result != null) {
+      return result;
+    }
+    Boolean inverseResult = checkValidityNoInverse(relation.inverse(), knownRelations);
+    if (inverseResult != null) {
+      return Boolean.valueOf(!inverseResult.booleanValue());
+    }
+    return null;
+  }
+
+  @Nullable
+  public static Boolean checkValidityNoInverse(SymbolicValueRelation relation, Collection<SymbolicValueRelation> knownRelations) {
+    SymbolicValueRelation symmetric = relation.symmetric();
+    for (SymbolicValueRelation otherRelation : knownRelations) {
+      Boolean validity = checkValidity(relation, otherRelation, knownRelations);
+      if (validity != null) {
+        return validity;
+      }
+      validity = checkValidity(symmetric, otherRelation, knownRelations);
+      if (validity != null) {
+        return validity;
+      }
+    }
+    return null;
+  }
+
+  @Nullable
+  private static Boolean checkValidity(SymbolicValueRelation rel1, SymbolicValueRelation rel2, Collection<SymbolicValueRelation> knownRelations) {
+    if (rel1.v1.equals(rel2.v1) && rel1.v2.equals(rel2.v2)) {
+      return rel1.consistentWith(rel2);
+    }
+    if (rel1.v2.equals(rel2.v1) && !rel1.v1.equals(rel2.v2)) {
+      SymbolicValueRelation combined = rel1.combineWith(rel2);
+      if (combined != null) {
+        List<SymbolicValueRelation> relations = new ArrayList<>(knownRelations);
+        relations.remove(rel2);
+        return checkValidity(combined, relations);
+      }
+    }
+    return null;
+  }
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder buffer = new StringBuilder();
+    buffer.append('(');
+    buffer.append(v1);
+    buffer.append(' ');
+    buffer.append(getOperand());
+    buffer.append(' ');
+    buffer.append(v2);
+    buffer.append(')');
+    return buffer.toString();
+  }
+
+  protected Boolean consistentWith(SymbolicValueRelation relation) {
+    return Boolean.valueOf(getClass().equals(relation.getClass()));
+  }
+
+  protected abstract String getOperand();
+
+  protected abstract SymbolicValueRelation inverse();
+
+  protected abstract SymbolicValueRelation symmetric();
+
+  @Nullable
+  protected abstract SymbolicValueRelation combineWith(SymbolicValueRelation relation);
```

Comment:
This one seems to be wrong in some ways in its implementation: there are 3 cases for this one : 
a == b, b ==c => a==b this one is correctly handled. 
a!=b, b!=c => unknown this one is correctly handled.  
a == b, b!=c => a!=c  this one is not correctly handled in this method (in either way) 

Things are working as expected in the end because this issue is adressed by verifying the relationship for the relationship and its inverse. This leads to extra computation because this can be adressed directly here.